### PR TITLE
TIP-1048: Upgrade all Doctrine packages

### PR DESCRIPTION
**Pull request by jjanvier**
Upgrade all Doctrine packages to the latest stable possible version.
I also set all packages that we didn't define before (even if they were installed). (Honestly, I don't like that, but as this is what we do for other libraries...)

Also, with this PR, **the PIM is installable with PHP 7.3**. It removes all the warnings and errors that came from Doctrine about: 

```
PHP Warning:  "continue" targeting switch is equivalent to "break". Did you mean to use "continue 2"? in /home/jjanvier/workspaces/phpstorm/akeneo/pim_master/pcd/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php on line 2636
```

I removed some specs that mocked `UnitOfWork`. Now Doctrine [returns a reference](https://github.com/doctrine/orm/blob/2.6/lib/Doctrine/ORM/UnitOfWork.php#L553) on `getEntityChangeSet` and [PHPSpec is not able to handle it](https://github.com/phpspec/prophecy/blob/master/src/Prophecy/Doubler/Generator/Node/MethodNode.php#L173).

EDIT: I tried to import/export 2500 products by CSV (Community Edition) and I didn't see any significative performance impacts. See comments for full results.

### CODE REVIEW

**Comment by jjanvier** [comment]
@BitOne should we merge that for PIM 3.1 or PIM 4.0?

**Comment by jjanvier** [comment]
Fixes 9623

**Comment by skeleton** [comment]
did you run a product import to check potential regression on perf?

**Comment by stefandoorn** [comment]
I'm a bit new to the project, but what's the reason not to use something like:

`"doctrine/annotations": "^1.6.0",`

Or:

`"doctrine/annotations": "~1.6.0",`

File: src/Akeneo/Pim/Structure/Bundle/Doctrine/ORM/Repository/InternalApi/AttributeOptionSearchableRepository.php [hunk 1]
```diff
@@ -57,21 +57,23 @@ public function findBySearch($search = null, array $options = [])
             ->andWhere('a.code = :attributeCode')
             ->setParameter('attributeCode', $options['identifier']);
 
-        if ($this->isAttributeAutoSorted($options['identifier']) && isset($options['catalogLocale'])) {
+        if (isset($options['identifier']) && $this->isAttributeAutoSorted($options['identifier']) && isset($options['catalogLocale'])) {
+
             $qb
-                ->addSelect('v.value AS HIDDEN')
+                ->addSelect('v.value AS HIDDEN value')
```
Comment (jjanvier) [inline src/Akeneo/Pim/Structure/Bundle/Doctrine/ORM/Repository/InternalApi/AttributeOptionSearchableRepository.php]: this was broken with a the Doctrine update

File: src/Akeneo/Pim/Structure/Bundle/Doctrine/ORM/Repository/InternalApi/AttributeOptionSearchableRepository.php [hunk 2]
```diff
@@ -57,21 +57,23 @@ public function findBySearch($search = null, array $options = [])
             ->andWhere('a.code = :attributeCode')
             ->setParameter('attributeCode', $options['identifier']);
 
-        if ($this->isAttributeAutoSorted($options['identifier']) && isset($options['catalogLocale'])) {
+        if (isset($options['identifier']) && $this->isAttributeAutoSorted($options['identifier']) && isset($options['catalogLocale'])) {
```
Comment (jjanvier) [inline src/Akeneo/Pim/Structure/Bundle/Doctrine/ORM/Repository/InternalApi/AttributeOptionSearchableRepository.php]: `isset($options['identifier']) ` was missing

**Comment by jjanvier** [comment]
## On master

### Export 2500 products

jjanvier:pcd$ time docker-compose exec fpm bin/console akeneo:batch:job productcsvexport --env=prod                                                                               
Export productcsvexport has been successfully executed.
docker-compose exec fpm bin/console akeneo:batch:job productcsvexport   1.16s user 0.16s system 1% cpu 1:14.72 total

### Import 2500 new products

jjanvier:pcd$ time docker-compose exec fpm bin/console akeneo:batch:job productcsvimport --env=prod                                                                               Import productcsvimport has been successfully executed.
docker-compose exec fpm bin/console akeneo:batch:job productcsvimport   1.07s user 0.10s system 0% cpu 5:38.57 total

 ## On this branch

### Export 2500 products

jjanvier:pcd$ time docker-compose exec fpm bin/console akeneo:batch:job productcsvexport --env=prod                                                                    
Export productcsvexport has been successfully executed.
docker-compose exec fpm bin/console akeneo:batch:job productcsvexport   1.09s user 0.13s system 1% cpu 1:16.84 total

### Import 2500 new products

jjanvier:pcd$ time docker-compose exec fpm bin/console akeneo:batch:job productcsvimport --env=prod                                                                    
Import productcsvimport has been successfully executed.
docker-compose exec fpm bin/console akeneo:batch:job productcsvimport   0.98s user 0.10s system 0% cpu 5:44.07 total

**Comment by jjanvier** [comment]
@stefandoorn  because we don't have a `.lock` file in the standard editions (the ones used for real life projects). Thus, we want to be sure about the versions of our dependencies.

**Comment by jjanvier** [comment]
the catalog configuration I used:

```yml
catalog:
    entities:
        channels:
            - code: ecommerce
              locales: [en_US, fr_FR, de_DE, es_ES, fr_BE]
              currencies: [USD, EUR]
            - code: print
              locales: [en_US, fr_FR, de_DE, es_ES, fr_BE]
              currencies: [USD, EUR]
            - code: web
              locales: [en_US, fr_FR, de_DE, es_ES, fr_BE]
              currencies: [USD, EUR]
        category-trees:
            -
                code: suppliers
                children: 30
                levels: 4
            -
                code: sales
                children: 30
                levels: 4
            -
                code: master
                children: 30
                levels: 4
        attribute-groups:
            count: 50
        attributes:
            count: 550
            percentage_of_localizable: 25
            percentage_of_scopable: 13
            percentage_of_localizable_and_scopable: 13
            percentage_of_useable_in_grid: 50
        families:
            count: 125
            attributes_count: 100
        products:
            count: 2500
            with_images: false
```

**Pull request merged**